libs/dmlf/src/local_vm_launcher.cpp

```diff
@@ -0,0 +1,163 @@
+  auto                   newProgram = std::make_shared<Program>();
+  auto errors = fetch::vm_modules::VMFactory::Compile(module_, files, *newProgram);
+
+  if (!errors.empty() && programErrorHandler_ != nullptr)
+  {
+    programErrorHandler_(name, errors);
+    return false;
+  }
+
+  programs_.emplace(std::move(name), std::move(newProgram));
+
+  return true;
+}
+bool LocalVmLauncher::HasProgram(std::string const &name) const
+{
+  return programs_.find(name) != programs_.end();
+}
+void LocalVmLauncher::AttachProgramErrorHandler(ProgramErrorHandler newHandler)
+{
+  programErrorHandler_ = newHandler;
+}
+
+bool LocalVmLauncher::CreateVM(std::string name)
+{
+  if (HasVM(name))
+  {
+    return false;
+  }
+  vms_.emplace(std::move(name), std::make_shared<VM>(module_.get()));
+  return true;
+}
+bool LocalVmLauncher::HasVM(std::string const &name) const
+{
+  return vms_.find(name) != vms_.end();
+}
+bool LocalVmLauncher::SetVmStdout(std::string const &vmName, VmOutputHandler &newHandler)
+{
+  auto it = vms_.find(vmName);
+  if (it == vms_.end())
+  {
+    return false;
+  }
+
+  auto vm = it->second;
+  vm->AttachOutputDevice(VM::STDOUT, newHandler);
+
+  return true;
+}
+
+bool LocalVmLauncher::CreateState(std::string name)
+{
+  if (HasState(name))
+  {
+    return false;
+  }
+  states_.emplace(std::move(name), std::make_shared<State>());
+  return true;
+}
+bool LocalVmLauncher::HasState(std::string const &name) const
+{
+  return states_.find(name) != states_.end();
+}
+bool LocalVmLauncher::CopyState(std::string const &srcName, std::string newName)
+{
+  if (!HasState(srcName) || HasState(newName))
+  {
+    return false;
+  }
+
+  states_.emplace(std::move(newName), std::make_shared<State>(states_[srcName]->DeepCopy()));
+
+  return true;
+}
+
+bool LocalVmLauncher::Execute(std::string const &programName, std::string const &vmName,
+                              std::string const &stateName, std::string const &entrypoint,
+                              Params const &params)
+{
+  if (!HasProgram(programName) || !HasVM(vmName) || !HasState(stateName))
+  {
+    return false;
+  }
+
+  auto &program = programs_[programName];
+  auto &vm      = vms_[vmName];
+  auto &state   = states_[stateName];
+
+  vm->SetIOObserver(*state);
+
+  fetch::vm::ParameterPack parameterPack(vm->registered_types());
+  std::for_each(params.cbegin(), params.cend(),
+                [&parameterPack](auto const &v) { parameterPack.AddSingle(v); });
+
+  std::string        runTimeError;
+  fetch::vm::Variant output;
+  auto thereWasAnError = vm->Execute(*program, entrypoint, runTimeError, output, parameterPack);
+
+  if (!runTimeError.empty() && executeErrorhandler_ != nullptr)
+  {
+    executeErrorhandler_(programName, vmName, stateName, runTimeError);
+    return false;
+  }
+
+  return thereWasAnError;
+}
+void LocalVmLauncher::AttachExecuteErrorHandler(ExecuteErrorHandler newHandler)
+{
+  executeErrorhandler_ = newHandler;
```

Comment:
I can't see from the review but presumably:

```suggestion
  executeErrorhandler_ = std::move(newHandler);
```